### ISSUE TITLE
Overlay fail but underlay works

### ISSUE THREAD

**Issue by LeoFCardoso** [qpdf/qpdf#904]
Hello, please consider this as a possible bug.

Overlay file: [overlay.pdf](https://github.com/qpdf/qpdf/files/10774243/overlay.pdf)
Source file: [source.pdf](https://github.com/qpdf/qpdf/files/10774245/source.pdf)

This generates a bad result (overlay looks inverted):
`qpdf --overlay ./overlay.pdf -- ./source.pdf ./result_overlay.pdf`

This generates a correct file:
`qpdf --underlay ./source.pdf -- ./overlay.pdf ./result_underlay.pdf`

Maybe related to https://github.com/qpdf/qpdf/issues/559

Related with https://github.com/LeoFCardoso/pdf2pdfocr/issues/40

**Comment by jberkenbilt** [comment]
I am not able to reproduce this with qpdf 11.2.0. Both files render identically (based on machine comparison, not visual inspection).

**Comment by LeoFCardoso** [comment]
Thank you for the fast reply.

Here are more technical details. I'm using macos.

```
macleonardo-2016:Downloads leonardo$ qpdf --version
qpdf version 11.2.0
Run qpdf --copyright to see copyright and license information.
macleonardo-2016:Downloads leonardo$ qpdf --overlay ./overlay.pdf -- ./source.pdf ./result_overlay.pdf
macleonardo-2016:Downloads leonardo$ qpdf --underlay ./source.pdf -- ./overlay.pdf ./result_underlay.pdf
macleonardo-2016:Downloads leonardo$ md5sum result_*.pdf
305058f51e04d28af0234669145d41e0  result_overlay.pdf
78997697199b35329dd83b69ad81035c  result_underlay.pdf
macleonardo-2016:Downloads leonardo$ uname -a
Darwin macleonardo-2016 21.6.0 Darwin Kernel Version 21.6.0: Mon Dec 19 20:44:01 PST 2022; root:xnu-8020.240.18~2/RELEASE_X86_64 x86_64
```
result_overlay: [result_overlay.pdf](https://github.com/qpdf/qpdf/files/10776811/result_overlay.pdf)
<img width="961" alt="overlay_print" src="https://user-images.githubusercontent.com/1037439/219946743-99842f54-5e18-4467-9bd1-b3ab5d8f0a74.png">

result_underlay: [result_underlay.pdf](https://github.com/qpdf/qpdf/files/10776838/result_underlay.pdf)
<img width="963" alt="underlay_print" src="https://user-images.githubusercontent.com/1037439/219946823-6ecbc458-fa6e-455d-a007-ec9b7ac55f3e.png">

**Comment by jberkenbilt** [comment]
There was an important piece of missing information here....overlay.pdf looks blank but contains hidden text. You have to actually select text to see the effect. I didn't get this from the description, but when I decoded the overlay file and looked at it, I saw that it was filled with hidden text. If I hit CTRL-a after performing your operations, then I can see it. Now I have enough information to look into this. Thanks.

**Comment by jberkenbilt** [comment]
This one is quite interesting. I appreciate your taking time to report it. I have fully analyzed this and understand what's going on.

There's an error with the source file that qpdf is not taking into consideration. It would be possible to change how qpdf does overlay and underlay to make it not susceptible to this problem. In a nutshell, PDF allows you to push graphics state onto a stack and restore it. qpdf and numerous other tools do this to isolate a block of code from changes made by other code in the content stream. The source file here modifies the state stack in an unbalanced way, which means that qpdf's isolation of the original page is incomplete. There would be a way for qpdf to isolate the original page in a stronger fashion. I'll look into that fix. It would cause a change to the way overlay/underlay is implemented for all files, but I think it would be good for underlay/overlay to work even when there are errors on the source page.

Here's a detailed explanation of what's going on.

In PDF content, the `q` operator pushes the current graphics state (including scaling, etc.) onto a stack, and the `Q` operator restores the old graphics state by popping it off of the stack. In this way, surrounding a block of content by `q`...`Q` isolates any changes it makes to the graphics state from other changes. The way qpdf's overlay and underlay feature works (and this would be true of pretty much any tool that is not a full PDF interpreter) is that it prepends `q` before the entire content and then appends `Q` to it, thus isolating the existing content. Then it adds a little code either before or after (depending on underlay or overlay) that draws the item being underlaid or overlaid.

The problem is that the original source has unbalanced `q` and `Q` operators. The original source does nothing but draw an image. Somewhat strangely, it does this by inverting the entire page, then drawing the image inverted on the page. This is most likely done by some software so that it can think of coordinates as 0,0 at the upper left, which is typical of image file formats, rather than 0,0 at the lower left, which is used by PDF and PostScript (and usually mathematical Cartesian coordinate systems).

The content stream from the original source is this, with comments added by me:

```
stream
1 0 0 -1 0 1658 cm    % flip the page about the bottom edge, then shift up by the height, resulting in vertical flip
q
1648 0 0 -1658 0 1658 cm   % draw the image the same way: flip about the bottom, then shift
/I6 Do
Q
q    % Note: unbalanced q operator
```

In the overlay case (the one that's broken), qpdf ends up with the following, again comments inserted by me

```
q  % qpdf adds this to isolate the original content stream
% original content stream begins here
1 0 0 -1 0 1658 cm
q
1648 0 0 -1658 0 1658 cm
/I6 Do
Q
q
% original content stream ends here
Q  % qpdf adds this to end isolation of original content stream
% qpdf adds this to draw the overlay
q
0.9999 0 0 0.9999 0.03951 0 cm
/Fx1 Do
Q
```

If you look at the `q` and `Q` operators, you can see the problem:

```
q   % BEGIN 1
1 0 0 -1 0 1658 cm
q   % BEGIN 2
1648 0 0 -1658 0 1658 cm
/I6 Do
Q  % END 2 (back to 1)
q  % BEGIN 2

Q % END 2 (back to 1)
q % BEGIN 2 -- changes from "1" are still in effect, hence the inversion of the page!
0.9999 0 0 0.9999 0.03951 0 cm
/Fx1 Do
Q % END 2
```

The new code that qpdf adds to draw the overlay is this `/Fx1 Do` where `/Fx1` is a "form xobject", which is a self-contained block of code that is more strongly isolated than is achieved with just `q` and `Q`. The fix would be to have qpdf put the original page in a form xobject as well as the overlay/underlay page. That would fully protect it from pages that are unbalanced in this way. The fact that underlaying the source works shows that this approach would work.

I'll take a look at the code and see how big a fix that would be.

**Comment by jberkenbilt** [comment]
Yes, I'll be able to fix this for 11.3.0, which I plan on releasing today if possible.

**Comment by johnwhitington** [comment]
>There's an error with the source file that qpdf is not taking into consideration. 

Note that, prior to ISO standardisation, unbalanced q/Q operators were considered perfectly kosher. In fact, the possibility was specifically mentioned in the PDF 1.7 standard, rather than just allowed by omission, if I remember correctly.
